File: testing/PIMS.Tests.Automation/PageObjects/SharedTeamMembers.cs
using OpenQA.Selenium;
using PIMS.Tests.Automation.Classes;

namespace PIMS.Tests.Automation.PageObjects
{
    public class SharedTeamMembers : PageObjectBase
    {
        private readonly By membersTeamSubtitle = By.XPath("//h2/div/div[contains(text(),'Team')]");
        private readonly By membersTeamAddAnotherMemberLink = By.CssSelector("button[data-testid='add-team-member']");
        private readonly By fileTeamMembersGroup = By.XPath("//div[contains(text(),'Team')]/parent::div/parent::h2/following-sibling::div/div[@class='py-3 row']");
        private readonly By membersTeamFirstMemberDeleteBttn = By.XPath("//div[contains(text(),'Team')]/parent::div/parent::h2/following-sibling::div/div[@class='py-3 row'][1]/div[3]/button");
        private readonly By membersTeamInvalidTeamMemberMessage = By.XPath("//div[contains(text(),'Select a team member')]");
        private readonly By membersTeamInvalidProfileMessage = By.XPath("//div[contains(text(),'Select a profile')]");

        //Acquisition File Confirmation Modal Elements
        private readonly By fileConfirmationModal = By.CssSelector("div[class='modal-content']");

        private SharedModals sharedModals;
        private SharedSelectContact sharedSelectContact;

        public SharedTeamMembers(IWebDriver webDriver) : base(webDriver)
        {
            sharedModals = new SharedModals(webDriver);
            sharedSelectContact = new SharedSelectContact(webDriver);
        }

        public void AddTeamMembers(TeamMember teamMember)
        {
            Wait();
            FocusAndClick(membersTeamAddAnotherMemberLink);

            Wait();
            var teamMemberIndex = webDriver.FindElements(fileTeamMembersGroup).Count() -1;

            WaitUntilVisible(By.CssSelector("select[id='input-team."+ teamMemberIndex +".contactTypeCode']"));
            ChooseSpecificSelectOption(By.CssSelector("select[id='input-team."+ teamMemberIndex +".contactTypeCode']"), teamMember.TeamMemberRole);
            FocusAndClick(By.CssSelector("div[data-testid='teamMemberRow["+ teamMemberIndex +"]'] div[data-testid='contact-input'] button[title='Select Contact']"));

            Wait();
            sharedSelectContact.SelectContact(teamMember.TeamMemberContactName, teamMember.TeamMemberContactType);

            Wait();
            if (webDriver.FindElements(By.Id("input-team."+ teamMemberIndex +".primaryContactId")).Count > 0)
                ChooseSpecificSelectOption(By.Id("input-team."+ teamMemberIndex +".primaryContactId"), teamMember.TeamMemberPrimaryContact);
        }

        public void AddMgmtTeamMembers(TeamMember teamMember)
        {
            Wait();
            FocusAndClick(membersTeamAddAnotherMemberLink);

            Wait();
            var teamMemberIndex = webDriver.FindElements(fileTeamMembersGroup).Count() -1;

            WaitUntilVisible(By.CssSelector("select[id='input-team."+ teamMemberIndex +".teamProfileTypeCode']"));
            ChooseSpecificSelectOption(By.CssSelector("select[id='input-team."+ teamMemberIndex +".teamProfileTypeCode']"), teamMember.TeamMemberRole);
            FocusAndClick(By.CssSelector("div[data-testid='teamMemberRow["+ teamMemberIndex +"]'] div[data-testid='contact-input'] button[title='Select Contact']"));

            Wait();
            sharedSelectContact.SelectContact(teamMember.TeamMemberContactName, teamMember.TeamMemberContactType);

            Wait();
            if (webDriver.FindElements(By.Id("input-team."+ teamMemberIndex +".primaryContactId")).Count > 0)
                ChooseSpecificSelectOption(By.Id("input-team."+ teamMemberIndex +".primaryContactId"), teamMember.TeamMemberPrimaryContact);
        }

        public void VerifyTeamMembersViewForm(List<TeamMember> teamMembers)
        {
            var index = 1;

            for (var i = 0; i < teamMembers.Count; i++)
            {
                AssertTrueContentEquals(By.XPath("//h2/div/div[contains(text(),'Team')]/parent::div/parent::h2/following-sibling::div/div[" + index + "]/div/label"), teamMembers[i].TeamMemberRole + ":");
                AssertTrueContentEquals(By.XPath("//h2/div/div[contains(text(),'Team')]/parent::div/parent::h2/following-sibling::div/div[" + index + "]/div/a"), teamMembers[i].TeamMemberContactName);

                if (teamMembers[i].TeamMemberPrimaryContact != "")
                {
                    index++;
                    AssertTrueContentEquals(By.XPath("//h2/div/div[contains(text(),'Team')]/parent::div/parent::h2/following-sibling::div/div[" + index + "]/div/a"), teamMembers[i].TeamMemberPrimaryContact);
                }

                index++;
            }
        }

        public void VerifyRequiredTeamMemberMessages()
        {
            //Add a new Team member form
            Wait();
            WaitUntilClickable(membersTeamAddAnotherMemberLink);
            webDriver.FindElement(membersTeamAddAnotherMemberLink).Click();

            //Verify that invalid team member message is displayed
            ChooseSpecificSelectOption(By.Id("input-team.0.contactTypeCode"), "Expropriation agent");
            webDriver.FindElement(membersTeamSubtitle).Click();
            AssertTrueIsDisplayed(membersTeamInvalidTeamMemberMessage);

            //verify that invalid profile message is displayed
            webDriver.FindElement(By.CssSelector("div[data-testid='contact-input'] button[title='Select Contact']")).Click();
            sharedSelectContact.SelectContact("Test", "");
            ChooseSpecificSelectOption(By.Id("input-team.0.contactTypeCode"), "Select profile...");
            webDriver.FindElement(membersTeamSubtitle).Click();
            AssertTrueIsDisplayed(membersTeamInvalidProfileMessage);
        }

        public void DeleteFirstStaffMember()
        {
            WaitUntilClickable(membersTeamFirstMemberDeleteBttn);
            webDriver.FindElement(membersTeamFirstMemberDeleteBttn).Click();

            WaitUntilVisible(fileConfirmationModal);
            Assert.Equal("Remove Team Member", sharedModals.ModalHeader());
            //Assert.Equal("Do you wish to remove this team member?", sharedModals.ModalContent());

            sharedModals.ModalClickOKBttn();
        }
    }
}

File: testing/playwright-testing/support/pages/SharedTeamMembers.js
const SharedSelectContact = require("./SharedSelectContacts");
const SharedModal = require("./SharedModal");
const { expect } = require("@playwright/test");

class SharedTeamMembers {
  constructor(page) {
    this.page = page;
    this.sharedSelectContact = new SharedSelectContact(this.page);
    this.sharedModal = new SharedModal(page);
  }

  async addTeamMembers(teamMember) {
    await this.page.getByTestId("add-team-member").click();

    const countMembersIdx =
      (await this.page.locator("div[data-testid^='teamMemberRow']").count()) -
      1;

    const contactTypeselects = this.page.locator(
      "select[id^='input-team.'][id$='.contactTypeCode']"
    );
    await contactTypeselects
      .nth(countMembersIdx)
      .selectOption({ label: teamMember.TeamMemberRole });

    await this.page.locator("button[title='Select Contact']").last().click();
    await this.sharedSelectContact.selectContact(
      teamMember.TeamMemberContactName,
      teamMember.TeamMemberContactType
    );

    if (
      await this.page
        .locator(`#input-team.${countMembersIdx}.primaryContactId`)
        .isVisible()
    ) {
      await this.page
        .locator(`#input-team.${countMembersIdx}.primaryContactId`)
        .selectOption({ label: teamMember.TeamMemberPrimaryContact });
    }
  }

  async addMgmtTeamMembers(teamMember) {
    await this.page.getByTestId("add-team-member").click();

    const countMembersIdx =
      (await this.page.locator("div[data-testid^='teamMemberRow']").count()) -
      1;
    const escaped = await this.page.evaluate(
      (str) => CSS.escape(str),
      `${countMembersIdx}`
    );
    const contactTypeselects = this.page.locator(
      "select[id^='input-team.'][id$='.teamProfileTypeCode']"
    );
    await contactTypeselects
      .nth(countMembersIdx)
      .selectOption({ label: teamMember.TeamMemberRole });

    await this.page.locator("button[title='Select Contact']").last().click();
    await this.sharedSelectContact.selectContact(
      teamMember.TeamMemberContactName,
      teamMember.TeamMemberContactType
    );

    if (
      await this.page
        .locator(`#input-team.${escaped}.primaryContactId`)
        .isVisible()
    ) {
      await this.page
        .locator(`#input-team.${escaped}.primaryContactId`)
        .selectOption({ label: teamMember.TeamMemberPrimaryContact });
    }
  }

  async verifyTeamMembersViewForm(teamMembers) {
    let index = 1;

    for (let i = 0; i < teamMembers.length; i++) {
      await expect(
        this.page
          .locator(
            `//div[@data-testid="teamMember[${i}]"]/preceding-sibling::div/label`
          )
          .textContent()
      ).toBe(teamMembers[i].TeamMemberRole + ":");
      await expect(
        this.page.locator(`div[data-testid='teamMember[${i}]'] span`)
      ).toBe(teamMembers[i].TeamMemberContactName);

      if (
        await this.page
          .locator(`div[data-testid='primaryContact[${i}]'] span`)
          .isVisible()
      ) {
        index++;
        await expect(
          this.page.locator(`div[data-testid='primaryContact[${i}]'] span`)
        ).toBe(teamMembers[i].TeamMemberPrimaryContact);
      }

      index++;
    }
  }

  async verifyRequiredTeamMemberMessages() {
    //Add a new Team member form
    await this.page.getByTestId("add-team-member").click();

    //Verify that invalid team member message is displayed
    await this.page
      .locator("#input-team.0.contactTypeCode")
      .selectOption({ label: "Expropriation agent" });
    await this.page.locator("//h2/div/div[contains(text(),'Team')]").click();
    await expect(this.page.locator("invalid-feedback").textContent()).toBe(
      "Select a team member"
    );

    //verify that invalid profile message is displayed
    await this.page
      .locator(
        "div[data-testid='contact-input'] button[title='Select Contact']"
      )
      .click();
    await this.sharedSelectContact.selectContact("Test", "");
    await this.page
      .locator("#input-team.0.contactTypeCode")
      .selectOption({ label: "Select profile..." });
    await this.page.locator("//h2/div/div[contains(text(),'Team')]").click();
    await expect(this.page.locator("invalid-feedback").textContent()).toBe(
      "Select a profile"
    );
  }

  async deleteFirstStaffMember() {
    await this.page.getByTestId("team.0.remove-button").click();

    await this.page
      .locator('div[class="modal-content"]')
      .waitFor({ state: "visible" });

    const actualHeaderTitle = await this.sharedModal.mainModalHeader();
    expect(actualHeaderTitle).toEqual("Remove Team Member");

    const actualModalContent = await this.sharedModal.mainModalContent();
    expect(actualModalContent).toEqual(
      "Do you wish to remove this team member?"
    );
    await this.sharedModal.mainModalClickOKBttn();
  }
}

module.exports = SharedTeamMembers;
